File: src/main/java/com/sf_lolitahag/panel/GamePanel.java
/*
Copyright (C) 2016 Koji Wakamiya
Copyright (C) 2016 Masanori Kojima
Released under the MIT license
http://opensource.org/licenses/mit-license.php
*/
package com.sf_lolitahag.panel;

import com.sf_lolitahag.Utils;
import com.sf_lolitahag.hanabi.FireworkManager;
import com.sf_lolitahag.motion.AbstractMotion;
import com.sf_lolitahag.motion.Furin;
import com.sf_lolitahag.motion.Human;
import com.sf_lolitahag.motion.Kitsune;
import java.awt.Graphics;
import java.awt.Image;
import java.util.ArrayList;
import java.util.List;
import javax.swing.Timer;

public class GamePanel extends AbstractPanel {

  private static final int PAINT_INTERVAL = 33;
  private static final String BACK = "back";
  private static final String ROOM = "room";
  private static final String SKY = "sky";
  private final Image back;
  private final Image room;
  private final Image sky;
  private final List<AbstractMotion> motions = new ArrayList<>();
  private Human human;
  private Kitsune kitsune;
  private FireworkManager fireworkManager;

  public GamePanel() {
    Class tmpClass = getClass();
    back = Utils.getImageFromResources(tmpClass, BACK);
    room = Utils.getImageFromResources(tmpClass, ROOM);
    sky = Utils.getImageFromResources(tmpClass, SKY);
    fireworkManager = new FireworkManager();
    initMotions();
    startRepaintTimer();
  }

  @Override
  public void onSpaceKeyPress() {
    human.onSpaceKeyPress();
  }

  @Override
  protected void paintComponent(final Graphics g) {
    g.drawImage(sky, 0, 0, null);
    g.drawImage(back, 0, 0, null);
    g.drawImage(room, 0, 0, null);

    motions.stream().filter(AbstractMotion::isShow)
        .forEach(m -> g.drawImage(m.getBodyImage(), m.getAxisX(), m.getAxisY(), null));

    fireworkManager.draw(g);
  }

  private void initMotions() {
    human = new Human(new Human.Callback() {
      @Override
      public void onStartOkosuMotion() {
        kitsune.setVisibility(false);
      }

      @Override
      public void onFinishOkosuMotion() {
        kitsune.startManzoku();
        kitsune.setVisibility(true);
      }
    });
    kitsune = new Kitsune(new Kitsune.Callback() {
      @Override
      public boolean isCheckHumanSleep() {
        return human.isSleepNow();
      }

      @Override
      public void onStartOkosu() {
        human.startOkosu();
      }
    });

    motions.add(new Furin());
    motions.add(human);
    motions.add(kitsune);
  }

  private void startRepaintTimer() {
    new Timer(PAINT_INTERVAL, (e) -> repaint()).start();
  }
}

File: src/main/kotlin/com/sf_lolitahag/panel/GamePanel.kt
/*
Copyright (C) 2016 Koji Wakamiya
Copyright (C) 2016 Masanori Kojima
Released under the MIT license
http://opensource.org/licenses/mit-license.php
*/
package com.sf_lolitahag.panel

import com.sf_lolitahag.Utils
import com.sf_lolitahag.hanabi.FireworkManager
import com.sf_lolitahag.motion.Furin
import com.sf_lolitahag.motion.Human
import com.sf_lolitahag.motion.Kitsune
import java.awt.Graphics
import javax.swing.JPanel
import javax.swing.Timer

class GamePanel : JPanel() {

    private val back by lazy { Utils.getImageFromResources(javaClass, BACK) }
    private val room by lazy { Utils.getImageFromResources(javaClass, ROOM) }
    private val sky by lazy { Utils.getImageFromResources(javaClass, SKY) }
    private val fireworkManager = FireworkManager()

    private val motions by lazy {
        listOf(Furin(), human, kitsune)
    }
    private val human by lazy {
        Human(object : Human.Callback {
            override fun onStartOkosuMotion() {
                kitsune.setVisibility(false)
            }

            override fun onFinishOkosuMotion() {
                kitsune.startManzoku()
                kitsune.setVisibility(true)
            }
        })
    }
    private val kitsune: Kitsune

    init {
        startRepaintTimer()

        kitsune = Kitsune(object : Kitsune.Callback {
            override fun isCheckHumanSleep(): Boolean {
                return human.isSleepNow
            }

            override fun onStartOkosu() {
                human.startOkosu()
            }
        })
    }

    fun onSpaceKeyPress() {
        human.onSpaceKeyPress()
    }

    override fun paintComponent(g: Graphics) {
        g.drawImage(sky, 0, 0, null)
        g.drawImage(back, 0, 0, null)
        g.drawImage(room, 0, 0, null)

        motions.filter {
            it.isShow
        }.forEach {
            g.drawImage(it.bodyImage, it.axisX, it.axisY, null)
        }

        fireworkManager.draw(g)
    }

    private fun startRepaintTimer() {
        Timer(PAINT_INTERVAL) { e -> repaint() }.start()
    }

    companion object {
        private const val PAINT_INTERVAL = 33
        private const val BACK = "back"
        private const val ROOM = "room"
        private const val SKY = "sky"
    }
}
